File: sdk/python/src/graft/modules/servo.py
"""GRAFT Servo Module"""

from __future__ import annotations

import struct
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graft.device import GraftDevice

from graft.core.constants import CMD


class GraftServo:
    def __init__(self, device: GraftDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def attach(self, pin: int, min_us: int = 544, max_us: int = 2400) -> None:
        payload = bytes([self._mod_id, 0x01, pin,
                         min_us & 0xFF, (min_us >> 8) & 0xFF,
                         max_us & 0xFF, (max_us >> 8) & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def write(self, angle: int) -> None:
        payload = bytes([self._mod_id, 0x02, angle & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def write_microseconds(self, us: int) -> None:
        payload = bytes([self._mod_id, 0x03, us & 0xFF, (us >> 8) & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def detach(self) -> None:
        payload = bytes([self._mod_id, 0x04])
        await self._device._send_command(CMD.MOD_CMD, payload)

File: sdk/go/modules/servo/servo.go
// Package servo provides a typed wrapper for the CONDUYT servo module.
package servo

import (
	"context"
	"encoding/binary"

	conduyt "github.com/virgilvox/conduyt/sdk/go"
)

// Servo provides typed control of a servo module.
type Servo struct {
	device   *conduyt.Device
	moduleID byte
}

// New creates a Servo wrapper for the given device and module ID.
func New(device *conduyt.Device, moduleID byte) *Servo {
	return &Servo{device: device, moduleID: moduleID}
}

// Attach attaches the servo to a pin.
func (s *Servo) Attach(ctx context.Context, pin byte, minUs, maxUs uint16) error {
	payload := make([]byte, 7)
	payload[0] = s.moduleID
	payload[1] = 0x01
	payload[2] = pin
	binary.LittleEndian.PutUint16(payload[3:5], minUs)
	binary.LittleEndian.PutUint16(payload[5:7], maxUs)
	return s.device.ModCmd(ctx, payload)
}

// Write writes an angle (0-180).
func (s *Servo) Write(ctx context.Context, angle byte) error {
	return s.device.ModCmd(ctx, []byte{s.moduleID, 0x02, angle})
}

// WriteMicroseconds writes a pulse width.
func (s *Servo) WriteMicroseconds(ctx context.Context, us uint16) error {
	payload := make([]byte, 4)
	payload[0] = s.moduleID
	payload[1] = 0x03
	binary.LittleEndian.PutUint16(payload[2:4], us)
	return s.device.ModCmd(ctx, payload)
}

// Detach releases the servo pin.
func (s *Servo) Detach(ctx context.Context) error {
	return s.device.ModCmd(ctx, []byte{s.moduleID, 0x04})
}
